lib/srv/alpnproxy/helpers_test.go

```diff
@@ -0,0 +1,217 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package alpnproxy
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+type Suite struct {
+	serverListener net.Listener
+	router         *Router
+	ca             *tlsca.CertAuthority
+}
+
+func NewSuite(t *testing.T) *Suite {
+	ca := mustGenSelfSignedCert(t)
+	pool := x509.NewCertPool()
+	pool.AddCert(ca.Cert)
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		l.Close()
+	})
+
+	router := NewRouter()
+	router.Add(HandlerDecs{
+		Protocols: []string{ProtocolDefault},
+		Handler: func(ctx context.Context, conn net.Conn) error {
+			t.Fatal("default handler called")
+			return nil
+		},
+	})
+
+	return &Suite{
+		ca:             ca,
+		serverListener: l,
+		router:         router,
+	}
+}
+
+func (s *Suite) GetServerAddress() string {
+	return s.serverListener.Addr().String()
+}
+
+func (s *Suite) GetCertPool() *x509.CertPool {
+	pool := x509.NewCertPool()
+	pool.AddCert(s.ca.Cert)
+	return pool
+}
+
+func (s *Suite) Start(t *testing.T) {
+	serverCert := mustGenCertSignedWithCA(t, s.ca)
+	proxyConfig := ProxyConfig{
+		Listener: s.serverListener,
+		TLSConfig: &tls.Config{
+			ClientAuth: tls.VerifyClientCertIfGiven,
+			ClientCAs:  s.GetCertPool(),
+			Certificates: []tls.Certificate{
+				serverCert,
+			},
+		},
+		Router: s.router,
+		Log:    logrus.New(),
+	}
+
+	svr, err := New(proxyConfig)
+	require.NoError(t, err)
+
+	syncC := make(chan struct{})
+	go func() {
+		close(syncC)
+		err := svr.Serve(context.Background())
+		require.NoError(t, err)
+	}()
+	<-syncC
```

Comment:
does `syncC` do anything here?

Reply:
It was suppose to make sure that the `svr.Serve` go-routine was scheduled. I will switch to retry calls in case of toggling tests.